pointmatcher/DataPointsFilters/utils/octree.hpp

```diff
@@ -153,457 +151,238 @@ Octree<T>& Octree<T>::operator=(Octree<T>&&o)
-		//Assign parent
-		octants[i]->parent = this;
-	}
 
-	return (!isLeaf() and ret);
-}
-
-template< typename T >
-bool Octree<T>::build_par(const DP& pts, DataContainer&& datas, BoundingBox && bb, T maxSizeByNode)
-{
-	static XYZ offsetTable[8] =
-		{
-			XYZ{-0.5, -0.5, -0.5},
-			XYZ{+0.5, -0.5, -0.5},
-			XYZ{-0.5, -0.5, +0.5},
-			XYZ{+0.5, -0.5, +0.5},
-			XYZ{-0.5, +0.5, -0.5},
-			XYZ{+0.5, +0.5, -0.5},
-			XYZ{-0.5, +0.5, +0.5},
-			XYZ{+0.5, +0.5, +0.5}
-		};
-	
-	//Check maxData count
-	if((bb.radius*2.0 <= maxSizeByNode) or (datas.size() <= 1))
-	{			
+	//Check stop condition
+	if((bb.radius*2.0 <= maxSizeByNode) or (datas.size() <= maxDataByNode))
+	{		
 		//insert data
 		data.insert(data.end(), 
 			std::make_move_iterator(datas.begin()), make_move_iterator(datas.end()));
 		return (isLeaf());
 	}
 	
-	//Assign bounding box
-	this->bb.center = bb.center;
-	this->bb.radius = bb.radius;
-	
 	//Split datas
 	const std::size_t nbData = datas.size();
 	
-	DataContainer sDatas[8];
-	for(size_t i=0; i<8; ++i)
+	DataContainer sDatas[nbCells];
+	for(size_t i=0; i<nbCells; ++i)
 		sDatas[i].reserve(nbData);
 		
 	for(auto&& d : datas)
-	{
-		//FIXME: Should be a generic conversion from DataPoint considering Data to XYZ
-		(sDatas[idx( TO_XYZ(pts,d) )]).emplace_back(d);
-	}
+		(sDatas[idx(pts, d)]).emplace_back(d);
 	
-	for(size_t i=0; i<8; ++i)
+	for(size_t i=0; i<nbCells; ++i)
 		sDatas[i].shrink_to_fit();
 	
 	//Compute new bounding boxes
-	BoundingBox boxes[8];
+	BoundingBox boxes[nbCells];
 	const T half_radius = this->bb.radius * 0.5;
-	for(size_t i=0; i<8; ++i)
+	for(size_t i=0; i<nbCells; ++i)
 	{
-		const XYZ offset = offsetTable[i] * this->bb.radius;
+		const Point offset = OctreeHelper<T,dim>::offsetTable[i] * this->bb.radius;
 		boxes[i].radius = half_radius;
 		boxes[i].center = this->bb.center + offset;
 	}
 	
-	//For each child build recursively	
-	std::vector<std::thread> threads;
+	//For each child build recursively
 	bool ret = true;
-	for(size_t i=0; i<8; ++i)
-	{
-		threads.push_back( std::thread( [maxSizeByNode, i, &pts, &sDatas, &boxes, this](){
-			this->octants[i] = new Octree<T>();
-			//Assign depth
-			this->octants[i]->depth = this->depth+1;	
-			//Assign parent
-			this->octants[i]->parent = this;
-			this->octants[i]->build(pts, std::move(sDatas[i]), std::move(boxes[i]), maxSizeByNode);	
-		}));
-	}
-	
-	// Wait for all thread to finish
-	for (std::thread & t : threads)
-		if (t.joinable()) t.join();
+	std::vector<std::future<void>> futures;
+
+	for(size_t i=0; i<nbCells; ++i)
+	{		
+		auto compute = [maxDataByNode, maxSizeByNode, i, &pts, &sDatas, &boxes, this](){
+				this->cells[i] = new Octree_<T,dim>();
+				//Assign depth
+				this->cells[i]->depth = this->depth+1;
+				//Assign parent
+				this->cells[i]->parent = this;
+				//next call is not parallelizable
+				this->cells[i]->build(pts, std::move(sDatas[i]), std::move(boxes[i]), maxDataByNode, maxSizeByNode, false);	
+			};
 		
+		if(not parallelBuild)
```

Comment:
Isn't that beautifully readable?

One last minor here: I recommend removing (almost blindly) outer most logic negations from if-conditions while exchanging the branch bodies iff the if statement has an else branch. 
This makes it even more readable (because you remove one operation from the users mind).

Reply:
Sure, I used to put the statement that occurs the most often in the _if_ statement, but indeed it's more readable for the user that way. I'll take that into account for future development!